### PR TITLE
macos: Enable notarization for macOS Catalina.

### PR DESCRIPTION
**Pull request by kanishk98**
---

Fixes an issue Catalina users were facing with Zulip's lack of notarization.

**You have tested this PR on:**
  - [ ] Windows
  - [ ] Linux/Ubuntu
  - [ ] macOS

### CODE REVIEW

File: scripts/notarize.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+exports.default = async function notarizing(context) {
+  const { electronPlatformName, appOutDir } = context;  
+  if (electronPlatformName !== 'darwin') {
+    return;
+  }
+
+  const appName = context.packager.appInfo.productFilename;
+
+  return await notarize({
+    appBundleId: 'com.yourcompany.yourAppId',
```
Comment (akashnimare) [inline scripts/notarize.js]: We the bundle ID in other places so we can re-use the same here.

Comment (akashnimare) [reply scripts/notarize.js]: You can import it from helpers.

**Comment by akashnimare** [comment]
@kanishk98 this is what I'm getting on running `npm run dist` -
```sh
➜  zulip-desktop git:(review-original-834) ✗ npm run dist

> zulip@4.0.1 dist /Users/akka/dev/zulipwork/zulip-desktop
> tsc && electron-builder

Configuring yargs through package.json is deprecated and will be removed in the next major release, please use the JS API instead.
Configuring yargs through package.json is deprecated and will be removed in the next major release, please use the JS API instead.
  • electron-builder version=20.40.2
  • loaded configuration file=package.json ("build" field)
  • effective config config=
                       directories:
                         output: dist
                         buildResources: build
                       afterSign: scripts/notarize.js
                       appId: org.zulip.zulip-electron
                       asar: true
                       asarUnpack:
                         - '**/*.node'
                       files:
                         - filter:
                             - '**/*'
                             - '!docs${/*}'
                             - '!node_modules/@paulcbetts/cld/deps/cld${/*}'
                       copyright: '©2019 Kandra Labs, Inc.'
                       mac:
                         category: public.app-category.productivity
                         darkModeSupport: true
                         artifactName: '${productName}-${version}-${arch}.${ext}'
                         hardenedRuntime: true
                         entitlements: build/entitlements.mac.plist
                         entitlementsInherit: build/entitlements.mac.plist
                         gatekeeperAssess: false
                       linux:
                         category: Chat;GNOME;GTK;Network;InstantMessaging
                         icon: build/icon.icns
                         packageCategory: GNOME;GTK;Network;InstantMessaging
                         description: Zulip Desktop Client for Linux
                         target:
                           - deb
                           - zip
                           - AppImage
                           - snap
                         maintainer: Akash Nimare <akash@zulipchat.com>
                         artifactName: '${productName}-${version}-${arch}.${ext}'
                       deb:
                         synopsis: Zulip Desktop App
                         afterInstall: ./scripts/debian-add-repo.sh
                         afterRemove: ./scripts/debian-uninstaller.sh
                       snap:
                         synopsis: Zulip Desktop App
                       dmg:
                         background: build/appdmg.png
                         icon: build/icon.icns
                         iconSize: 100
                         sign: false
                         contents:
                           - x: 380
                             'y': 280
                             type: link
                             path: /Applications
                           - x: 110
                             'y': 280
                             type: file
                         window:
                           width: 500
                           height: 500
                       win:
                         target:
                           - target: nsis-web
                             arch:
                               - x64
                               - ia32
                           - target: msi
                             arch:
                               - x64
                               - ia32
                         icon: build/icon.ico
                         artifactName: '${productName}-Web-Setup-${version}.${ext}'
                         publisherName: 'Kandra Labs, Inc.'
                       nsis:
                         allowToChangeInstallationDirectory: true
                         oneClick: false
                         perMachine: false
                       
  • two package.json structure is used devPackageFile=/Users/akka/dev/zulipwork/zulip-desktop/package.json appPackageFile=/Users/akka/dev/zulipwork/zulip-desktop/app/package.json
Error: Configuration is invalid.
 - configuration.dmg should be one of these:
   object { artifactName?, background?, backgroundColor?, contents?, format?, icon?, iconSize?, iconTextSize?, internetEnabled?, publish?, title?, window? } | {
     "type": "null"
   }
   macOS DMG options.
   
   Details:
    * configuration.dmg has an unknown property 'sign'. These properties are valid:
      object { artifactName?, background?, backgroundColor?, contents?, format?, icon?, iconSize?, iconTextSize?, internetEnabled?, publish?, title?, window? }
    * configuration.dmg should be null:
      {
        "type": "null"
      }
 - configuration.mac should be one of these:
   object { appId?, artifactName?, asar?, asarUnpack?, binaries?, bundleShortVersion?, bundleVersion?, category?, compression?, cscInstallerKeyPassword?, cscInstallerLink?, cscKeyPassword?, cscLink?, darkModeSupport?, detectUpdateChannel?, electronLanguages?, electronUpdaterCompatibility?, entitlements?, entitlementsInherit?, extendInfo?, extraDistFiles?, extraFiles?, extraResources?, fileAssociations?, files?, forceCodeSigning?, generateUpdatesFilesForAllChannels?, helperBundleId?, icon?, identity?, minimumSystemVersion?, protocols?, provisioningProfile?, publish?, releaseInfo?, requirements?, target?, type? } | {
     "type": "null"
   }
   Options related to how build macOS targets.
   
   Details:
    * configuration.mac has an unknown property 'hardenedRuntime'. These properties are valid:
      object { appId?, artifactName?, asar?, asarUnpack?, binaries?, bundleShortVersion?, bundleVersion?, category?, compression?, cscInstallerKeyPassword?, cscInstallerLink?, cscKeyPassword?, cscLink?, darkModeSupport?, detectUpdateChannel?, electronLanguages?, electronUpdaterCompatibility?, entitlements?, entitlementsInherit?, extendInfo?, extraDistFiles?, extraFiles?, extraResources?, fileAssociations?, files?, forceCodeSigning?, generateUpdatesFilesForAllChannels?, helperBundleId?, icon?, identity?, minimumSystemVersion?, protocols?, provisioningProfile?, publish?, releaseInfo?, requirements?, target?, type? }
    * configuration.mac has an unknown property 'gatekeeperAssess'. These properties are valid:
      object { appId?, artifactName?, asar?, asarUnpack?, binaries?, bundleShortVersion?, bundleVersion?, category?, compression?, cscInstallerKeyPassword?, cscInstallerLink?, cscKeyPassword?, cscLink?, darkModeSupport?, detectUpdateChannel?, electronLanguages?, electronUpdaterCompatibility?, entitlements?, entitlementsInherit?, extendInfo?, extraDistFiles?, extraFiles?, extraResources?, fileAssociations?, files?, forceCodeSigning?, generateUpdatesFilesForAllChannels?, helperBundleId?, icon?, identity?, minimumSystemVersion?, protocols?, provisioningProfile?, publish?, releaseInfo?, requirements?, target?, type? }
    * configuration.mac should be null:
      {
        "type": "null"
      }

How to fix:
1. Open https://electron.build/configuration/configuration
2. Search the option name on the page.
  * Not found? The option was deprecated or not exists (check spelling).
  * Found? Check that the option in the appropriate place. e.g. "title" only in the "dmg", not in the root.

    at validateConfig (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/read-config-file/src/main.ts:133:19)
    at <anonymous>
From previous event:
    at _validateConfig2 (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/out/util/config.js:339:27)
    at validateConfig (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/out/util/config.js:302:27)
    at /Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/packager.ts:336:11
    at Generator.next (<anonymous>)
From previous event:
    at Packager._build (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/packager.ts:335:133)
    at /Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/packager.ts:331:23
    at Generator.next (<anonymous>)
    at runCallback (timers.js:810:20)
    at tryOnImmediate (timers.js:768:5)
    at processImmediate [as _immediateCallback] (timers.js:745:5)
From previous event:
    at Packager.build (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/packager.ts:288:14)
    at build (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/index.ts:59:28)
    at build (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/electron-builder/src/builder.ts:228:10)
    at then (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/electron-builder/src/cli/cli.ts:46:19)
    at <anonymous>
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! zulip@4.0.1 dist: `tsc && electron-builder`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the zulip@4.0.1 dist script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/akka/.npm/_logs/2019-10-29T20_14_36_190Z-debug.log
➜  zulip-desktop git:(review-original-834) ✗ npm run dist > log.txt     
Configuring yargs through package.json is deprecated and will be removed in the next major release, please use the JS API instead.
Configuring yargs through package.json is deprecated and will be removed in the next major release, please use the JS API instead.
Error: Configuration is invalid.
 - configuration.dmg should be one of these:
   object { artifactName?, background?, backgroundColor?, contents?, format?, icon?, iconSize?, iconTextSize?, internetEnabled?, publish?, title?, window? } | {
     "type": "null"
   }
   macOS DMG options.
   
   Details:
    * configuration.dmg has an unknown property 'sign'. These properties are valid:
      object { artifactName?, background?, backgroundColor?, contents?, format?, icon?, iconSize?, iconTextSize?, internetEnabled?, publish?, title?, window? }
    * configuration.dmg should be null:
      {
        "type": "null"
      }
 - configuration.mac should be one of these:
   object { appId?, artifactName?, asar?, asarUnpack?, binaries?, bundleShortVersion?, bundleVersion?, category?, compression?, cscInstallerKeyPassword?, cscInstallerLink?, cscKeyPassword?, cscLink?, darkModeSupport?, detectUpdateChannel?, electronLanguages?, electronUpdaterCompatibility?, entitlements?, entitlementsInherit?, extendInfo?, extraDistFiles?, extraFiles?, extraResources?, fileAssociations?, files?, forceCodeSigning?, generateUpdatesFilesForAllChannels?, helperBundleId?, icon?, identity?, minimumSystemVersion?, protocols?, provisioningProfile?, publish?, releaseInfo?, requirements?, target?, type? } | {
     "type": "null"
   }
   Options related to how build macOS targets.
   
   Details:
    * configuration.mac has an unknown property 'hardenedRuntime'. These properties are valid:
      object { appId?, artifactName?, asar?, asarUnpack?, binaries?, bundleShortVersion?, bundleVersion?, category?, compression?, cscInstallerKeyPassword?, cscInstallerLink?, cscKeyPassword?, cscLink?, darkModeSupport?, detectUpdateChannel?, electronLanguages?, electronUpdaterCompatibility?, entitlements?, entitlementsInherit?, extendInfo?, extraDistFiles?, extraFiles?, extraResources?, fileAssociations?, files?, forceCodeSigning?, generateUpdatesFilesForAllChannels?, helperBundleId?, icon?, identity?, minimumSystemVersion?, protocols?, provisioningProfile?, publish?, releaseInfo?, requirements?, target?, type? }
    * configuration.mac has an unknown property 'gatekeeperAssess'. These properties are valid:
      object { appId?, artifactName?, asar?, asarUnpack?, binaries?, bundleShortVersion?, bundleVersion?, category?, compression?, cscInstallerKeyPassword?, cscInstallerLink?, cscKeyPassword?, cscLink?, darkModeSupport?, detectUpdateChannel?, electronLanguages?, electronUpdaterCompatibility?, entitlements?, entitlementsInherit?, extendInfo?, extraDistFiles?, extraFiles?, extraResources?, fileAssociations?, files?, forceCodeSigning?, generateUpdatesFilesForAllChannels?, helperBundleId?, icon?, identity?, minimumSystemVersion?, protocols?, provisioningProfile?, publish?, releaseInfo?, requirements?, target?, type? }
    * configuration.mac should be null:
      {
        "type": "null"
      }

How to fix:
1. Open https://electron.build/configuration/configuration
2. Search the option name on the page.
  * Not found? The option was deprecated or not exists (check spelling).
  * Found? Check that the option in the appropriate place. e.g. "title" only in the "dmg", not in the root.

    at validateConfig (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/read-config-file/src/main.ts:133:19)
    at <anonymous>
From previous event:
    at _validateConfig2 (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/out/util/config.js:339:27)
    at validateConfig (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/out/util/config.js:302:27)
    at /Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/packager.ts:336:11
    at Generator.next (<anonymous>)
From previous event:
    at Packager._build (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/packager.ts:335:133)
    at /Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/packager.ts:331:23
    at Generator.next (<anonymous>)
    at runCallback (timers.js:810:20)
    at tryOnImmediate (timers.js:768:5)
    at processImmediate [as _immediateCallback] (timers.js:745:5)
From previous event:
    at Packager.build (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/packager.ts:288:14)
    at build (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/index.ts:59:28)
    at build (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/electron-builder/src/builder.ts:228:10)
    at then (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/electron-builder/src/cli/cli.ts:46:19)
    at <anonymous>
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! zulip@4.0.1 dist: `tsc && electron-builder`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the zulip@4.0.1 dist script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/akka/.npm/_logs/2019-10-29T20_15_47_913Z-debug.log

```

**Comment by akashnimare** [comment]
Tried building got this 
```sh
TypeError: Cannot read property 'getPath' of undefined
    at Object.<anonymous> (/Users/akka/dev/zulipwork/zulip-desktop/app/renderer/js/utils/default-util.js:12:22)
    at Module._compile (module.js:653:30)
    at Object.Module._extensions..js (module.js:664:10)
    at Module.load (module.js:566:32)
    at tryModuleLoad (module.js:506:12)
    at Function.Module._load (module.js:498:3)
    at Module.require (module.js:597:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/akka/dev/zulipwork/zulip-desktop/app/renderer/js/utils/logger-util.js:3:24)
    at Module._compile (module.js:653:30)
    at Object.Module._extensions..js (module.js:664:10)
    at Module.load (module.js:566:32)
    at tryModuleLoad (module.js:506:12)
    at Function.Module._load (module.js:498:3)
    at Module.require (module.js:597:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/akka/dev/zulipwork/zulip-desktop/app/renderer/js/notification/helpers.js:4:16)
    at Module._compile (module.js:653:30)
    at Object.Module._extensions..js (module.js:664:10)
    at Module.load (module.js:566:32)
    at tryModuleLoad (module.js:506:12)
    at Function.Module._load (module.js:498:3)
From previous event:
    at MacPackager.doPack (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/platformPackager.ts:167:165)
    at /Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/macPackager.ts:90:63
    at Generator.next (<anonymous>)
From previous event:
    at MacPackager.pack (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/macPackager.ts:82:95)
    at /Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/packager.ts:430:24
    at Generator.next (<anonymous>)
    at xfs.stat (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/fs-extra/lib/mkdirs/mkdirs.js:56:16)
    at callback (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/fs-extra/node_modules/graceful-fs/polyfills.js:289:20)
    at FSReqWrap.oncomplete (fs.js:153:5)
From previous event:
    at Packager.doBuild (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/packager.ts:396:24)
    at /Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/packager.ts:366:57
    at Generator.next (<anonymous>)
    at /Users/akka/dev/zulipwork/zulip-desktop/node_modules/fs-extra/node_modules/graceful-fs/graceful-fs.js:111:16
    at /Users/akka/dev/zulipwork/zulip-desktop/node_modules/graceful-fs/graceful-fs.js:43:10
    at /Users/akka/dev/zulipwork/zulip-desktop/node_modules/fs-extra/node_modules/graceful-fs/graceful-fs.js:45:10
    at FSReqWrap.oncomplete (fs.js:135:15)
From previous event:
    at Packager._build (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/packager.ts:335:133)
    at /Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/packager.ts:331:23
    at Generator.next (<anonymous>)
    at runCallback (timers.js:810:20)
    at tryOnImmediate (timers.js:768:5)
    at processImmediate [as _immediateCallback] (timers.js:745:5)
From previous event:
    at Packager.build (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/packager.ts:288:14)
    at build (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/app-builder-lib/src/index.ts:59:28)
    at build (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/electron-builder/src/builder.ts:228:10)
    at then (/Users/akka/dev/zulipwork/zulip-desktop/node_modules/electron-builder/src/cli/cli.ts:49:19)
    at <anonymous>
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! zulip@4.0.1 dist: `tsc && electron-builder`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the zulip@4.0.1 dist script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/akka/.npm/_logs/2019-11-01T19_50_02_055Z-debug.log


```

**Comment by kanishk98** [comment]
@akashnimare looks like `app` is not defined in `default-util`, which is pretty weird - we haven't touched this file in the PR and have also accounted for both main and renderer processes accessing this module. 

Can you check what the `process.type` is and if `electron` is being imported as expected? I don't think there's a problem with our logic there.

**Comment by akashnimare** [comment]
@kanishk98 not sure why, I'm still getting this while building -
```sh

TypeError: Cannot read property 'getPath' of undefined
    at Object.<anonymous> (/Users/akash/dev/zulipwork/zulip-desktop/app/renderer/js/utils/default-util.js:12:22)
```

**Comment by akashnimare** [comment]
> @akashnimare looks like `app` is not defined in `default-util`, which is pretty weird - we haven't touched this file in the PR and have also accounted for both main and renderer processes accessing this module.
> 
> Can you check what the `process.type` is and if `electron` is being imported as expected? I don't think there's a problem with our logic there.

Let me check what's wrong here.

**Comment by akashnimare** [comment]
This might work -
```
import fs = require('fs');

let setupCompleted = false;

import electron = require('electron');

const { app } = electron.remote;

const zulipDir = app.getPath('userData');

```

I think `default.ts` should be in renderer process.

**Comment by akashnimare** [comment]
This has something to do with the ts migration I think. `process.type` says `browser`. @vsvipul any idea what could be wrong here?

File: package.json [hunk 2]
```diff
@@ -43,6 +43,7 @@
     "test"
   ],
   "build": {
+    "afterSign": "scripts/notarize.js",
```
Comment (akashnimare) [inline package.json]: This should be `./scripts/notarize.js` ?

**Pull request merged**